# EDA - Analisis exploratorio de datos

- Analisis de dataset originales
- Preprocesamiento
    - Limpieza de dataset
    - Armado de datasets para entrenamiento
        - Muestras totales (unitarias, por segundo y por minuto)
        - Muestras por especimen (unitarias, por segundo y por minuto)
- Analisis de datasets resultantes

In [1]:
# Imports

import pandas as pd
import seaborn as sns
import numpy as np


Dependiendo del kernel de python que estemos usando, esta bueno chequear si estamos corriendo en la maquina local, el contenedor de docker con jupyter u otro server remoto. 

In [2]:
!whoami
!pwd

jovyan
/home/jovyan


Si estamos corriendo con el contenedor que se crea con este repo, tenemos que movernos a la ubicación del repo, en este caso usando el comando magico %cd.

In [3]:
%cd meleros/notebooks/

/home/jovyan/meleros/notebooks


## Lectura de datasets originales

Partimos de 2 datasets originales, primero vamos a revisar la forma de ambos programaticamente:
- Head: primeras lineas
- Shape: Tamaño del dataframe

In [4]:
dataset1_path = "../data/0-cruda/estudio3.csv"
dataset1_crudo_dataframe = pd.read_csv(dataset1_path)

dataset2_path = "../data/0-cruda/tabla-resumen.csv"
dataset2_crudo_dataframe = pd.read_csv(dataset2_path)


In [5]:
dataset1_crudo_dataframe.head()

UE   Nombre    Sexo Etapa     Hora  Tiempo       X       Y       Z Estado  \
0   6  Matilda  Hembra    A1  20:00.0       1  0.0000 -0.5469  0.8125      I   
1   6  Matilda  Hembra    A1  20:00.1       1 -0.0156 -0.5469  0.8281      I   
2   6  Matilda  Hembra    A1  20:00.2       1  0.0000 -0.5469  0.7969      I   
3   6  Matilda  Hembra    A1  20:00.3       1 -0.0156 -0.5469  0.8125      I   
4   6  Matilda  Hembra    A1  20:00.4       1  0.0000 -0.5469  0.8125      I   

  Comportamiento      ODBA  
0              D  0.014043  
1              D  0.020277  
2              D  0.029643  
3              D  0.011963  
4              D  0.014043

In [6]:
dataset2_crudo_dataframe.head()

UE    Nombre    Sexo      Fecha     Hora Estado Comportamiento Comport 2  \
0   1  Chaqueña  Hembra  8/22/2019  25:28.0      A              E         E   
1   1  Chaqueña  Hembra  8/22/2019  25:28.1      A              E         E   
2   1  Chaqueña  Hembra  8/22/2019  25:28.2      A              E         E   
3   1  Chaqueña  Hembra  8/22/2019  25:28.3      A              E         E   
4   1  Chaqueña  Hembra  8/22/2019  25:28.4      A              E         E   

  Posición       x  ...  x prom/3seg  y prom/3seg  z prom/3seg     dif x  \
0        C  0.1250  ...     0.117717       0.1948     0.914077  0.007283   
1        C  0.1094  ...          NaN          NaN          NaN -0.008317   
2        C  0.1094  ...          NaN          NaN          NaN -0.008317   
3        C  0.1094  ...          NaN          NaN          NaN -0.008317   
4        C  0.1719  ...          NaN          NaN          NaN  0.054183   

   dif y      dif z  abs dif x  abs dif y   abs dif z       ODBA  
0  0.0708 -0.023477   0.007283      0.0708    0.023477  0.101560  
1  0.0708 -0.039077   0.008317      0.0708    0.039077  0.118193  
2  0.0708 -0.007777   0.008317      0.0708    0.007777  0.086893  
3  0.0552  0.007823   0.008317      0.0552    0.007823  0.071340  
4  0.0396  0.007823   0.054183      0.0396    0.007823  0.101607  

[5 rows x 22 columns]

In [19]:
# Revisar nombres de columnas
dataset1_crudo_dataframe.columns.to_list()

['UE',
 'Nombre',
 'Sexo',
 'Etapa',
 'Hora',
 'Tiempo',
 'X',
 'Y',
 'Z',
 'Estado',
 'Comportamiento',
 'ODBA']

In [20]:
dataset2_crudo_dataframe.columns.to_list()

['UE',
 'Nombre',
 'Sexo',
 'Fecha',
 'Hora',
 'Estado',
 'Comportamiento',
 'Comport 2',
 'Posición',
 'x',
 'y',
 'z',
 'x prom/3seg',
 'y prom/3seg',
 'z prom/3seg',
 'dif x',
 'dif y ',
 'dif z',
 'abs dif x',
 'abs dif y ',
 'abs dif z ',
 'ODBA']

In [9]:
dataset1_crudo_dataframe.shape


(466560, 12)

In [10]:
dataset2_crudo_dataframe.shape

(119030, 22)

In [11]:
dataset1_crudo_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466560 entries, 0 to 466559
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   UE              466560 non-null  int64  
 1   Nombre          466560 non-null  object 
 2   Sexo            466560 non-null  object 
 3   Etapa           466560 non-null  object 
 4   Hora            466560 non-null  object 
 5   Tiempo          466560 non-null  int64  
 6   X               466560 non-null  float64
 7   Y               466560 non-null  float64
 8   Z               466560 non-null  float64
 9   Estado          466559 non-null  object 
 10  Comportamiento  466559 non-null  object 
 11  ODBA            466560 non-null  float64
dtypes: float64(4), int64(2), object(6)
memory usage: 42.7+ MB


In [12]:
dataset2_crudo_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119030 entries, 0 to 119029
Data columns (total 22 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   UE              119030 non-null  int64  
 1   Nombre          119030 non-null  object 
 2   Sexo            119030 non-null  object 
 3   Fecha           119030 non-null  object 
 4   Hora            119030 non-null  object 
 5   Estado          119030 non-null  object 
 6   Comportamiento  119020 non-null  object 
 7   Comport 2       119020 non-null  object 
 8   Posición        119030 non-null  object 
 9   x               119030 non-null  float64
 10  y               119030 non-null  float64
 11  z               119030 non-null  float64
 12  x prom/3seg     3968 non-null    float64
 13  y prom/3seg     3968 non-null    float64
 14  z prom/3seg     3968 non-null    float64
 15  dif x           119030 non-null  float64
 16  dif y           119030 non-null  float64
 17  dif z     

In [13]:
dataset1_crudo_dataframe.describe()

UE         Tiempo              X              Y  \
count  466560.00000  466560.000000  466560.000000  466560.000000   
mean        5.50000     144.500000       0.205897      -0.197764   
std         2.81366      83.138027       0.307815       0.502086   
min         2.00000       1.000000      -1.500000      -2.875000   
25%         3.00000      72.750000       0.015600      -0.593800   
50%         5.00000     144.500000       0.234400      -0.203100   
75%         8.00000     216.250000       0.390600       0.140600   
max        10.00000     288.000000       1.750000       1.578100   

                   Z           ODBA  
count  466560.000000  466560.000000  
mean        0.540138       0.076477  
std         0.494675       0.136346  
min        -3.468800       0.000047  
25%         0.359400       0.017673  
50%         0.750000       0.027037  
75%         0.890600       0.051567  
max         5.250000       8.062487

In [14]:
dataset2_crudo_dataframe.describe()

UE              x              y              z  \
count  119030.000000  119030.000000  119030.000000  119030.000000   
mean        5.502310       0.278723      -0.023097       0.669473   
std         2.904318       0.422830       0.336086       0.348745   
min         1.000000      -2.000000      -2.125000      -7.296900   
25%         3.000000       0.015600      -0.281300       0.484400   
50%         6.000000       0.203100      -0.031300       0.812500   
75%         8.000000       0.625000       0.218800       0.921900   
max        10.000000       2.328100       3.234400       5.046900   

       x prom/3seg  y prom/3seg  z prom/3seg         dif x        dif y   \
count  3968.000000  3968.000000  3968.000000  1.190300e+05  1.190300e+05   
mean      0.278663    -0.023176     0.669728  1.097118e-18  2.355132e-19   
std       0.408476     0.313718     0.330369  1.096109e-01  1.209401e-01   
min      -0.994280    -1.018233    -0.691680 -1.748943e+00 -2.248437e+00   
25%       0.040109    -0.268242     0.498581 -3.226667e-02 -4.008667e-02   
50%       0.204443    -0.037527     0.822150  5.200000e-04  0.000000e+00   
75%       0.613554     0.212777     0.908344  3.128000e-02  4.014583e-02   
max       1.012520     0.945320     1.059917  1.992150e+00  3.110963e+00   

              dif z      abs dif x     abs dif y      abs dif z   \
count  1.190300e+05  119030.000000  119030.000000  119030.000000   
mean   1.271771e-18       0.064500       0.076805       0.063634   
std    1.129497e-01       0.088624       0.093421       0.093319   
min   -8.158363e+00       0.000000       0.000000       0.000000   
25%   -3.175667e-02       0.010900       0.012027       0.013053   
50%    4.333333e-05       0.031778       0.040120       0.032327   
75%    3.285333e-02       0.081786       0.108900       0.077103   
max    3.986983e+00       1.992150       3.110963       8.158363   

                ODBA  
count  119030.000000  
mean        0.204939  
std         0.216879  
min         0.001063  
25%         0.054710  
50%         0.135943  
75%         0.282347  
max        12.413090

In [15]:
# Chequear valores nulos dataset 1
dataset1_crudo_dataframe.isnull().sum()

UE                0
Nombre            0
Sexo              0
Etapa             0
Hora              0
Tiempo            0
X                 0
Y                 0
Z                 0
Estado            1
Comportamiento    1
ODBA              0
dtype: int64

In [16]:
# Chequear valores nulos dataset 2
dataset2_crudo_dataframe.isnull().sum()

UE                     0
Nombre                 0
Sexo                   0
Fecha                  0
Hora                   0
Estado                 0
Comportamiento        10
Comport 2             10
Posición               0
x                      0
y                      0
z                      0
x prom/3seg       115062
y prom/3seg       115062
z prom/3seg       115062
dif x                  0
dif y                  0
dif z                  0
abs dif x              0
abs dif y              0
abs dif z              0
ODBA                   0
dtype: int64

In [17]:
# Chequear valores unicos dataset 1
dataset1_crudo_dataframe.nunique()

UE                     6
Nombre                 6
Sexo                   2
Etapa                  3
Hora                1083
Tiempo               288
X                    158
Y                    158
Z                    206
Estado                 2
Comportamiento        12
ODBA              116897
dtype: int64

In [18]:
# Chequear valores unicos dataset 2
dataset2_crudo_dataframe.nunique()

UE                   10
Nombre               10
Sexo                  2
Fecha                 5
Hora              35700
Estado                2
Comportamiento       10
Comport 2             7
Posición              9
x                   151
y                   152
z                   179
x prom/3seg        3368
y prom/3seg        3444
z prom/3seg        2830
dif x             25492
dif y             28505
dif z             24145
abs dif x         18447
abs dif y         20023
abs dif z         18152
ODBA              57669
dtype: int64

### Notas

#### Dataset 1 - estudio3

- Pareciera ser el original crudo
- Mas muestras: (466560, 12)
- Menos columnas
    - Datos de interes: Nombre, Sexo, X, Y, Z,  Comportamiento (salida) y ODBA.
    - 12 tipos de comportamiento de salida

#### Dataset 2 - tabla-resumen

- Pareciera ser el dataset curado para el estudio con excel
- Menos muestras: (119030, 22)
- Mas columnas
    - Tiene un segundo comportamiento y promedios de los últimos 3 segundos de las lecturas del acelerometro. 
    - 10 tipos de comportamientos de salida

A priori pareciera ser que el dataset 1 es el mejor para representado. Para verificar, se revisan las columnas de salida y el numero de especímenes y muestras de ambos datasets.

In [25]:
# Verificar valores de la columna Comportamiento para ambos datasets
dataset1_crudo_dataframe['Comportamiento'].unique().tolist()

['D', 'E', 'AL', 'O', 'IM', 'AA', 'NV', 'L', 'LD', 'LA', 'LI', 'LR', nan]

In [26]:
dataset2_crudo_dataframe['Comportamiento'].unique().tolist()

['E', 'TD', 'IM', 'L', 'AA', 'TA', nan, 'OTROS', 'I', 'AL', 'A']

In [27]:
dataset1_crudo_dataframe['Nombre'].unique().tolist()

['Matilda', 'Maíz', 'Franco', 'Isabella', 'Garrita', 'Salvaje']

In [28]:
dataset2_crudo_dataframe['Nombre'].unique().tolist()

['Chaqueña',
 'Franco',
 'Garrita',
 'Inti',
 'Isabella',
 'Luz',
 'Maíz',
 'Matilda',
 'Salvaje',
 'Tucumano']

In [29]:
# Numero de muestras por nombre en dataset 1
dataset1_crudo_dataframe['Nombre'].value_counts()

Nombre
Matilda     77760
Maíz        77760
Franco      77760
Isabella    77760
Garrita     77760
Salvaje     77760
Name: count, dtype: int64

In [30]:
dataset2_crudo_dataframe['Nombre'].value_counts()

Nombre
Chaqueña    14600
Luz         13920
Inti        13080
Maíz        13040
Matilda     12400
Franco      12270
Tucumano    11830
Salvaje     11650
Garrita     10650
Isabella     5590
Name: count, dtype: int64

## Analisis detallado

### Dataset 1

- Muestras por especimen
- Muestras por comportamiento
- Muestras por sexo
- Muestras por especimen por comportamiento

In [ ]:
# Seaborn 